src/Views/MainView.vala

```diff
@@ -109,13 +94,69 @@ public class Sideload.MainView : AbstractView {
         });
     }
 
-    public void display_details (string? size, bool extra_repo) {
+    public void display_bundle_details (string size, bool extra_repo) {
+        download_size_label.label = _("Install size may be up to %s").printf (size);
+
+        if (extra_repo) {
+            var updates_icon = new Gtk.Image.from_icon_name ("system-software-update-symbolic", Gtk.IconSize.BUTTON);
+            updates_icon.valign = Gtk.Align.START;
+
+            unowned Gtk.StyleContext updates_context = updates_icon.get_style_context ();
+            updates_context.add_class (Granite.STYLE_CLASS_ACCENT);
+            updates_context.add_class ("orange");
+
+            var updates_label = new Gtk.Label (_("Updates to this app will not be reviewed"));
+            updates_label.selectable = true;
+            updates_label.max_width_chars = 50;
+            updates_label.wrap = true;
+            updates_label.xalign = 0;
+
+            var repo_icon = new Gtk.Image.from_icon_name ("system-software-install-symbolic", Gtk.IconSize.BUTTON);
+            repo_icon.valign = Gtk.Align.START;
+
+            unowned Gtk.StyleContext repo_context = repo_icon.get_style_context ();
+            repo_context.add_class (Granite.STYLE_CLASS_ACCENT);
+            repo_context.add_class ("purple");
+
+            var appstore_name = ((Sideload.Application) GLib.Application.get_default ()).get_appstore_name ();
+
+            var repo_label = new Gtk.Label (_("Other apps from this distributor may appear in %s").printf (appstore_name));
+            repo_label.selectable = true;
+            repo_label.max_width_chars = 50;
+            repo_label.wrap = true;
+            repo_label.xalign = 0;
+
+            details_grid.attach (updates_icon, 0, 1);
+            details_grid.attach (updates_label, 1, 1);
+            details_grid.attach (repo_icon, 0, 2);
+            details_grid.attach (repo_label, 1, 2);
+        }
+
+        details_stack.visible_child_name = "details";
+        show_all ();
+    }
+
+    public void display_ref_details (string? size, bool extra_repo) {
         if (size != null) {
             download_size_label.label = _("Download size may be up to %s").printf (size);
         } else {
             download_size_label.label = _("Unknown download size");
         }
 
+        var updates_icon = new Gtk.Image.from_icon_name ("system-software-update-symbolic", Gtk.IconSize.BUTTON);
```

Comment:
You should construct `updates_icon` and `updates_label` in the class constructor as they're exactly the same in both cases and then just attach them at the relevant position depending on the type of dialog you're building.

Reply:
Bundles only have updates if you specify a repository for that, otherwise, they have no updates. it don't make sense to show that if there's no repository to be added.

Reply:
I mean that you write exactly the same code to construct the widgets twice (line 146-157 and line 101-112). Just run this once in the constructor of the dialog, and only attach the widgets to the grid if you need them.